File: Semestru 4/MPP/lab/.net/lab3/lab3/Networking/protocol/Worker.cs
using Model.app.domain;
using Networking.rpcprotocol;
using Services.services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;

namespace Networking.protocol
{
	public class Worker : IObserver
	{
		private TcpClient Client;
		private IService Service;

		private NetworkStream Stream;
		private IFormatter Formatter;
		private volatile bool Connected;

		public Worker(IService service, TcpClient client)
		{
			this.Client = client;
			this.Service = service;
			try
			{
				this.Stream = client.GetStream();
				Formatter = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();
				Connected = true;
			} catch (Exception e)
			{
				Console.WriteLine("Error: " + e.Message);
			}
		}

		public void ChildAdded(Child child)
		{
			Response<int> response = new Response<int>() { Type = ResponseType.CHILD_ADDED, Entity = child };
			this.SendResponse(response);
		}

		public void EventAdded(Event event_)
		{
			Response<int> response = new Response<int>() { Type = ResponseType.EVENT_ADDED, Entity = event_ };
			this.SendResponse(response);
		}

		public void SignupAdded(Signup signup)
		{
			Response<int> response = new Response<int>() { Type = ResponseType.SIGNUP_ADDED, Entities = new List<Entity<int>> { signup.Child, signup.Event } };
			this.SendResponse(response);
		}

		public Response<int> HandleRequest(Request<int> request)
		{
			Response<int> response = null;
			switch(request.Type)
			{
				case RequestType.LOGIN:
				{
					LoginInfo login = (LoginInfo)request.Entity!;
					lock(this.Service)
					{
						if (this.Service.Login(login, this))
							response = new Response<int>() { Type = ResponseType.OK, Entity = login };
						else response = new Response<int>() { Type = ResponseType.ERROR, ErrorMessage = "Error!" };
					}
					break;
				}
				case RequestType.GET_LOGIN:
				{
					LoginInfo? login;
					lock (this.Service) {
						login = this.Service.GetLoginByUsername(request.Message!);
					}
					if (login == null)
						response = new Response<int>() { Type = ResponseType.ERROR, ErrorMessage = "Login not found!" };
					else response = new Response<int>() { Type = ResponseType.LOGIN_RECEIVED, Entity = login };
					break;
				}
				case RequestType.LOGOUT:
				{
					LoginInfo login = (LoginInfo)request.Entity!;
					lock (this.Service)
					{
						this.Service.Logout(login, this);
					}
					this.Connected = false;
					response = new Response<int>() { Type = ResponseType.OK };
					break;
				}
				case RequestType.ADD_CHILD:
				{
					Child child = (Child)request.Entity!;
					lock (this.Service)
					{
						response = new Response<int>() { Type = ResponseType.CHILD_ADDED, Entity = this.Service.AddChild(child.Name, child.Cnp) };
					}
					break;
				}
				case RequestType.ADD_EVENT:
				{
					Event event_ = (Event)request.Entity!;
					lock (this.Service)
					{
						response = new Response<int>() { Type = ResponseType.EVENT_ADDED, Entity = this.Service.AddEvent(event_.Name, event_.MinAge, event_.MaxAge) };
					}
					break;
				}
				case RequestType.ADD_SIGNUP:
				{
					Signup? signup;
					lock (this.Service)
					{
						signup = this.Service.AddSignup((Child)request.Entities!.ElementAt(0), (Event)request.Entities!.ElementAt(1))!;
					}
					if (signup == null)
						response = new Response<int>() { Type = ResponseType.ERROR, ErrorMessage = "Error!" };
					else response = new Response<int>() { Type = ResponseType.SIGNUP_ADDED, Entities = new List<Entity<int>> {signup.Child, signup.Event}};
					break;
				}
				case RequestType.GET_CHILDREN:
				{
					lock (this.Service)
					{
						response = new Response<int>() { Type = ResponseType.CHILDREN_RECEIVED, Entities = this.Service.GetAllChildren() };
					}
					break;
				}
				case RequestType.GET_EVENTS:
				{
					lock (this.Service)
					{
						response = new Response<int>() { Type = ResponseType.EVENTS_RECEIVED, Entities = this.Service.GetAllEvents() };
					}
					break;
				}
				case RequestType.GET_SIGNUPS:
				{
					lock (this.Service)
					{
						response = new Response<int>()
						{
							Type = ResponseType.SIGNUPS_RECEIVED,
							Entities = this.Service.GetAllSignups()
											.SelectMany(signup => new Entity<int>[] { signup.Child, signup.Event }).ToList()
						};
					}
					break;
				}
			}
			return response!;
		}

		private void SendResponse(Response<int> response)
		{
			Console.WriteLine($"Sending response {response}");
			lock (Stream)
			{
				Formatter.Serialize(Stream, response);
				Stream.Flush();
			}
		}

		public virtual void run()
		{
			while(this.Connected)
			{
				try
				{
					Request<int> request = (Request<int>)Formatter.Deserialize(Stream);
					Response<int> response = HandleRequest(request);
					if (response != null)
						this.SendResponse(response);
				} catch (Exception e)
				{
					Console.WriteLine("Error: " + e.Message);
					this.CloseConnection();
					break;
				}
			}
			this.CloseConnection();
		}

		private void CloseConnection()
		{
			this.Connected = false;
			Stream.Close();
			Client.Close();
		}
	}
}

File: Semestru 4/MPP/lab/java/lab3/Network/src/main/java/org/example/protocol/Worker.java
package org.example.protocol;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.domain.Child;
import org.example.domain.Event;
import org.example.domain.LoginInfo;
import org.example.domain.Signup;
import org.example.services.IObserver;
import org.example.services.IService;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class Worker implements Runnable, IObserver {
    private final IService server;
    private final Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;
    private static Logger logger = LogManager.getLogger(Worker.class);

    public Worker(IService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output=new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input=new ObjectInputStream(connection.getInputStream());
            connected=true;
        } catch (IOException e) {
            logger.error(e);
            logger.error(e.getStackTrace());
        }
    }

    public void run() {
        logger.debug("Worker thread started for client: {}", connection.getRemoteSocketAddress());
        while(connected){
            try {
                Object request=input.readObject();
                logger.debug("Received request from client: {}", request);
                Response response=handleRequest((Request)request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException|ClassNotFoundException e) {
                logger.error(e.getMessage());
                connected = false;
                e.printStackTrace();
            }
            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                logger.error(e);
                logger.error(e.getStackTrace());
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error {}",e);
        }
    }

    private static Response okResponse() { return new Response().setType(ResponseType.OK);}

    private Response handleRequest(Request request) {
        Response response = null;
        switch (request.getType()){
            case LOGIN: {
                var login = (LoginInfo)request.getEntity();
                var actual = server.GetByUsername(login.GetUsername());
                if (login.GetPassword().equals(actual.GetPassword())) {
                    server.login(actual, this);
                    response = okResponse();
                }
                else return new Response().setType(ResponseType.ERROR).setMessage("Invalid username or password");
                break;
            }
            case ADD_CHILD: {
                Child child = (Child) request.getEntity();
                response = new Response().setType(ResponseType.CHILD_ADDED).setEntity(server.AddChild(child.GetName(), child.GetCNP()));
                break;
            }
            case ADD_EVENT: {
                Event event = (Event) request.getEntity();
                response = new Response().setType(ResponseType.EVENT_ADDED).setEntity(server.AddEvent(event.GetName(), event.GetMinAge(), event.GetMaxAge()));
                break;
            }
            case ADD_SIGNUP: {
                Signup signup = (Signup) request.getEntity();
                response = new Response().setType(ResponseType.SIGNUP_ADDED).setEntity(server.AddSignup(signup.GetChild(), signup.GetEvent()));
                break;
            }
            case GET_CHILDREN: {
                response = new Response().setType(ResponseType.CHILDREN_RECEIVED).setEntities(StreamSupport.stream(server.GetAllChildren().spliterator(), false)
                                    .collect(Collectors.toList()));
                break;
            }
            case GET_EVENTS: {
                response = new Response().setType(ResponseType.EVENTS_RECEIVED).setEntities(StreamSupport.stream(server.GetAllEvents().spliterator(), false)
                        .collect(Collectors.toList()));
                break;
            }
            case GET_SIGNUPS: {
                response = new Response().setType(ResponseType.SIGNUPS_RECEIVED).setEntities(StreamSupport.stream(server.GetAllSignups().spliterator(), false)
                        .collect(Collectors.toList()));
                break;
            }
        }
        return response;
    }


    private void sendResponse(Response response) throws IOException{
        logger.debug("sending response {}", response);
        synchronized (output) {
            output.writeObject(response);
            output.flush();
        }
    }

    @Override
    public void childAdded(Child child) {
        var response = new Response().setType(ResponseType.CHILD_ADDED).setEntity(child);
        logger.debug("Child added {}", child.GetName());
        try {
            sendResponse(response);
        } catch (IOException e) {
            logger.error(e);
        }
    }

    @Override
    public void eventAdded(Event event) {
        var response = new Response().setType(ResponseType.EVENT_ADDED).setEntity(event);
        logger.debug("Event added {}", event.GetName());
        try {
            sendResponse(response);
        } catch (IOException e) {
            logger.error(e);
        }
    }

    @Override
    public void signupAdded(Signup signup) {
        var response = new Response().setType(ResponseType.SIGNUP_ADDED).setEntity(signup);
        logger.debug("Signup added {}", signup.GetChild().GetName() + " " + signup.GetEvent().GetName());
        try {
            sendResponse(response);
        } catch (IOException e) {
            logger.error(e);
        }
    }
}
